File: spacetimedb/src/systems/research.ts
import { RESEARCH_DT } from '../../../shared/constants.ts';
import {
  UPGRADE_DEFS,
  setTech,
  type Tech as TechT,
} from '../../../shared/research.ts';
import { spacetimedb } from '../schema/db.ts';
import { researchTimer } from '../schema/tables.ts';
import { scheduleRefs } from '../schema/schedule_refs.ts';

// Research progress — runs every RESEARCH_TICK_MS. Advances each in-flight tech by
// one tick of its researchTime; on completion it flips the owner's player.techMask
// bit (so combat math reads a single number) and marks the row done. Deterministic:
// progress is a fixed fraction per tick, no clocks/random. Done rows are kept so
// the HUD can show a tech as complete; the techMask is the authority for combat.
export const researchSystem = spacetimedb.reducer(
  { timer: researchTimer.rowType },
  (ctx) => {
    for (const r of [...ctx.db.research.iter()]) {
      if (r.done) continue;
      const up = UPGRADE_DEFS[r.tech as TechT];
      if (!up) continue;
      // researchTime is in seconds; a single tick adds DT/time of the bar.
      const step = up.researchTime > 0 ? RESEARCH_DT / up.researchTime : 1;
      const progress = r.progress + step;
      if (progress < 1) {
        ctx.db.research.researchId.update({ ...r, progress });
        continue;
      }
      // Completed: stamp the bit onto the owner's mask, then mark the row done.
      const p = ctx.db.player.identity.find(r.owner);
      if (p) {
        ctx.db.player.identity.update({
          ...p,
          techMask: setTech(p.techMask, r.tech as TechT),
        });
      }
      ctx.db.research.researchId.update({ ...r, progress: 1, done: true });
    }
  }
);

scheduleRefs.researchSystem = researchSystem;

File: saladin-bevy/crates/protocol/src/systems/research.rs
use crate::MatchStatuses;
use crate::components::{MatchId, Player, Research};
use bevy_ecs::prelude::*;
use bevy_platform::collections::HashMap;
use saladin_sim::{Fx, RESEARCH_DT, Tech, set_tech, upgrade_def};

/// Research progress — runs every research tick (1 s). Advances each in-flight
/// tech by one tick of its research time; on completion flips the owner's
/// `tech_mask` bit (combat reads one number) and marks the row done. Ported from
/// the SpacetimeDB `researchSystem` reducer.
pub fn research(
    statuses: Res<MatchStatuses>,
    mut q_research: Query<(&mut Research, &MatchId)>,
    mut q_players: Query<(Entity, &mut Player)>,
) {
    let player_ent: HashMap<u64, Entity> = q_players.iter().map(|(e, p)| (p.player_id, e)).collect();

    for (mut r, mid) in &mut q_research {
        if r.done || !statuses.simulates(mid.0) {
            continue;
        }
        let Some(tech) = Tech::from_u8(r.tech) else { continue };
        let up = upgrade_def(tech);
        let step = if up.research_time > Fx::ZERO { RESEARCH_DT / up.research_time } else { Fx::ONE };
        let progress = r.progress + step;
        if progress < Fx::ONE {
            r.progress = progress;
            continue;
        }
        if let Some(&pe) = player_ent.get(&r.owner) {
            if let Ok((_, mut p)) = q_players.get_mut(pe) {
                p.tech_mask = set_tech(p.tech_mask, tech);
            }
        }
        r.progress = Fx::ONE;
        r.done = true;
    }
}
